AraEvent/AraEventCalibrator.cxx

```diff
@@ -871,18 +901,89 @@ void AraEventCalibrator::calibrateEvent(UsefulAtriStationEvent *theEvent, AraCal
         loadAtriPedestals(thisStationId);
     }
 
-    //Step one is loop over the blocks
+    // 3rd step. Converts DAQ data format to Electronic channel format
+    // Bin trimming and Pedestal subtraction are now in the seperate functions
+    DAQtoEleCh(theEvent, voltMapIt, timeMapIt, sampleList, capArrayList);
+
+    // From here, each function is independent. So, You can call the functiions at any step
+    // As a default, prioritizing the functions that reduce the number of samples
+    
+    // 4th step. Erase first block that currupted by trigger
+    // Apply conditioner function here for the correct ZeroMean() and VoltCal() performance
+    if(hasTrimFirstBlock(calType)) {
+        hasTrimFirstBlk = TrimFirstBlock(theEvent, voltMapIt, timeMapIt, sampleList, capArrayList);
+    }    
+
+    // 5th step. Trims samples using fAtriSampleIndex table
+    // A trimmed sampleList will be used for calling the right pedestal subtraction samples and the conversion factor samples
+    if(AraCalType::hasBinWidthCalib(calType)){ 
+        BinWidthCalib(theEvent, voltMapIt, timeMapIt, sampleList, capArrayList, hasTrimFirstBlk);    
+    }
+    
+    // 6th step. Pedestal subtraction
+    if(hasPedestalSubtraction(calType)) {
+        PedSubtraction(theEvent, voltMapIt, sampleList);
+    }
+    
+    // 7th step. Common mode
+    if(hasCommonMode(calType)) {
+        CommonMode(theEvent, voltMapIt);
+    }
+
+    // 8th step. Inverts only RF channels = 0,4,8 in A3
+    // Apply conditioner function here for correct VoltCal() performance
+    if (hasInvertA3Chans(calType) && thisStationId ==3) {
+        InvertA3Chans(theEvent, voltMapIt, thisStationId);
+    }
+
+    // 9th step. Zeroing ADC WF by subtracting mean
+    // RJN change 13-Feb-2013
+    // Now do zero mean,   
+    // THM added 12-Feb-2014
+    // After zeroMean (and only then!!) do voltage calibration,
+    // KAH added 09-15-2020
+    // first, need to do voltage calibration. if you zero-mean the block beforehand, the ADC values get screwed up and the resulting waveform looks bad.
+    // each sample needs to be individually calibrated. This requires knowing the dda, the channel, the block and the sample%64. 
+    // The below code loops over each of these.
+    // the previous version of the code did the zero mean before the votlage calibration. If there are issues with A2 or A3 calibration, try switching the order back.
+    // MK added 27-09-2021
+    // Adds back just for A2 and A3. Dealing with outlier events
+    if(hasADCZeroMean(calType) && thisStationId != 5) {
```

Comment:
This step is unclear to me. Are you zero meaning before or after ADC conversion?

Reply:
I'm zero meaning before and after ADC conversion. so two times.
Zero meaning before the conversion is for outlier event(example in [slide 6~7](https://aradocs.wipac.wisc.edu/0024/002464/001/20210927_data_analysis_pedestal_and_calibration_issues_part2.pdf)) and slide 8 shows even though we use the repeder pedestal some of WFs are not nicely centering in zero. The best solution I can have is... for now is applying 'zero meaning before the conversion'. I think this was also the original setting before Kaeli apply the modification.
Zero meaning after the conversion is mainly for A3 data. Even though WF is centering in zero before ADC conversion, voltage results are centering in zero ([slide 9](https://aradocs.wipac.wisc.edu/0024/002473/001/20211004_data_analysis_AraRoot_calibration_code_updates.pdf)). So, I have to apply 2nd zero meaning for now.

We could remove the outlier event by some sort of filtering method and also perform recalibration to get a better conversion factor, But I think that is for future improvement.

Reply:
Maybe I can put the slide link in the comment section:)

Reply:
I'm not sure A5 has this kind of outlier event. So, I'm not applying 'Zero meaning before conversion' just for the A5 condition. 
I will put more descriptions in the file!

Reply:
Great, thank you. This looks good now.